contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -48,134 +49,248 @@ Enable user namespace support in a kubernetes cluster so that workloads that wor
-This support is already present (currently it seems broken, will be fixed) in Kubernetes as an experimental functionality, which can be enabled using `feature-gates=”ExperimentalHostUserNamespaceDefaulting=true”`.
-If Pod-Security-Policy is configured to allow the following to be requested by a  pod, host user-namespace will be enabled for the container:
-  - host namespaces (pid, ipc, net)
-  - non-namespaced capabilities (mknod, sys_time, sys_module)
-  - the pod contains a privileged container or using host path volumes.
-  - https://github.com/kubernetes/kubernetes/commit/d0d78f478ce0fb9d5e121db3b7c6993b482af82c#diff-a53fa76e941e0bdaee26dcbc435ad2ffR437 introduced via https://github.com/kubernetes/kubernetes/commit/d0d78f478ce0fb9d5e121db3b7c6993b482af82c.
+* **RunAsUser**: UID to run the entrypoint of the container process. After usernamespace support in kubelet, if Kubelet discovers that usernamespace remapping is enabled at runtime, this will be the UID in the container usernamespace. In such a case, if no mapping(corresponding host UID) is found in the mapping range, pod will be failed admission at Kubelet. If kubelet discovers that usernamespace remapping is not supported at runtime, no validation check will be performed. 
+* **RunAsGroup**: same as **RunAsUser** but for gid mapping ranges.
+* **RunAsNonRoot**: Today, kubelet fails to start the container if image is found to be having UID 0. After usernamespace support in kubelet, kubelet should ignore this check if runtime is found to be running with usernamespace remapping enabled. Kubelet should verify image UID only when it is learned that usernamespace remapping is either not supported or not enabled at runtime.
+* **FsGroup and SupplementalGroups**: FsGroup and SupplementalGroups will continue to work after usernamespace support is Kubelet if runtime either does not support usernamespaces or support is not enabled. In case usernamespace support is discovered by kubelet as enabled at runtime, FsGroups/SupplementalGroups will work only if supported by the runtime (CRI implementation). One implementation, for example, could be where runtime(CRI implementation) will add a triplet in `/proc/<container-pid>/gid_map` for the FsGroup/Supplemental GIDs. This will be a 1:1 mapping between host usernamespace group id and container usernamespace group id i.e container ID and host ID will be same. Example:
 
-- **Beta:** Explicit API to request host user-namespace in pod spec
-          This is being targeted under Beta graduation plans.
+    If pod has `FsGroup: 2000` and runtime's default gid mappings are `containerid 0,  hostid 100000, size 1000`, runtime should be adding two entries in the `/proc/<container-pid>/gid_map`:
+    - [0 100000 1000]  // From GID 0 to 1000 in the container are mapped with GID range 100k to 100k+1000 on host
+    - [2000 2000 1]    // GID 2000 in container usernamespace is mapped to GID 2000 is host usernamespace
```

Comment:
What if there's overlap? E.g. what if it was `FsGroup: 500` in this example?

Reply:
Will fail at kubelet admission